Tools/clinic/clinic.py

```diff
@@ -136,6 +138,24 @@ def text_accumulator() -> TextAccumulator:
     text, append, output = _text_accumulator()
     return TextAccumulator(append, output)
 
+
+class ClinicError(Exception):
+    def __init__(
+        self,
+        message: str,
+        /,
+        *,
+        lineno: int | None = None,
+        filename: str | None = None
+    ):
+        super().__init__(message)
+        self.lineno = lineno
+        self.filename = filename
+
+
+class ClinicWarning(Warning): ...
```

Comment:
Style nit: the convention (as far as I understand it) is that stub files, and other places where you'd never _expect_ to find a body (abstract methods, methods on protocols, methods decorated with `@overload`) should use `...` for empty bodies. But for places where you'd usually _expect_ to find some kind of a body (any other kind of a function in a non-stub file; any class definition in a non-stub file) should generally be explicit about the empty body by using `pass`:

```suggestion
class ClinicWarning(Warning): pass
```